assets/ca/ca.querytranslator.js

```diff
@@ -0,0 +1,387 @@
+     */
+	escapeValue = function (value) {
+		return value.replace(/([\-\+&\|!\(\)\{}\[\]\^"~\*\?:\\])/, '\\$1');
+	};
+
+	/**
+	 * Convert a set of rules from the jQuery query builder into a CA search query.  Performs the inverse operation to
+	 * `convertSearchQueryToQueryBuilderRules`.
+	 * @param {Object} ruleSet
+	 * @returns {String}
+	 */
+	caUI.convertQueryBuilderRuleSetToSearchQuery = function (ruleSet) {
+		if (ruleSet.condition && ruleSet.rules) {
+			return '(' + $.map(ruleSet.rules, caUI.convertQueryBuilderRuleSetToSearchQuery).join(' ' + ruleSet.condition + ' ') + ')';
+		}
+		if (ruleSet.operator && ruleSet.field) {
+			// Escape value to allow special characters
+			var negation = ruleSet.operator.match(/not_/),
+				prefix = ruleSet.field + (negation ? ':-' : ':');
+			switch (negation ? ruleSet.operator.replace('not_', '') : ruleSet.operator) {
+				case 'equal':
+					return prefix + '"' + escapeValue(ruleSet.value) + '"';
+				case 'in':
+					return prefix + '(' + escapeValue(ruleSet.value) + ')';
+				case 'between':
+					return prefix + '[' + escapeValue(ruleSet.value[0]) + ' TO ' + escapeValue(ruleSet.value[1]) + ']';
+				case 'begins_with':
+					return prefix + '"' + escapeValue(ruleSet.value) + '*"';
+				case 'contains':
+					return prefix + '"*' + escapeValue(ruleSet.value) + '*"';
+				case 'ends_with':
+					return prefix + '"*' + escapeValue(ruleSet.value) + '"';
+				case 'is_empty':
+				case 'is_null':
+					return prefix + '*';
+			}
+			return ruleSet.field + ':' + ruleSet.value;
+		}
+		return '';
+	};
+
+	/**
+	 * Retrieve the list of tokens from the given query string.
+	 * @param {String} query
+	 * @return {Array}
+     */
+	getTokenList = function (query) {
+		var token,
+			tokens = [],
+			queryArray = query.trim().split('');
+		while (token = shiftToken(queryArray)) {
+			tokens.push(token);
+		}
+		while (tokens.length >= 2 && tokens[0].type === TOKEN_LPAREN && tokens[tokens.length - 1].type === TOKEN_RPAREN) {
+			tokens = tokens.slice(1, tokens.length - 1);
+		}
+		return tokens;
+	};
+
+	/**
+	 * Retrieve the next token from the `queryArray`, which is destructively processed.
+	 * @param {Array} queryArray
+	 * @returns {Object|undefined}
+	 */
+	shiftToken = function (queryArray) {
+		var character, token,
+			quoted = false,
+			escaped = false,
+			end = false;
+		// End condition.
+		if (queryArray.length === 0) {
+			return undefined;
+		}
+		// Inspect the first character to determine the type of token.
+		character = queryArray.shift();
+		switch (character) {
+			// Single-character tokens.
+			case '(':
+				token = { type: TOKEN_LPAREN };
+				end = true;
+				break;
+			case ')':
+				token = { type: TOKEN_RPAREN };
+				end = true;
+				break;
+			case ':':
+				token = { type: TOKEN_COLON };
+				end = true;
+				break;
+			case '-':
+				token = { type: TOKEN_NEGATION };
+				end = true;
+				break;
+			// Collapse whitespace.
+			case ' ':
+				while (queryArray[0] === ' ') {
```

Comment:
I'm not sure about using a text editor, that's kind of what the query builder is for.

However it has led me to thinking about the fact that the query is deleted from the text field if it's invalid, and the reasons why it can't be like that.  For example you might be half way through writing the query and click on a menu item to check something in a new tab, and the query will be wiped.  Or you might just have written something that the query builder doesn't understand but is valid search query syntax.

I can change "show/hide query builder" to "enable/disable query builder", so the query builder is only engaged with the query when it is expanded (and at the time of expansion, update the query builder).  I can also leave the search query intact when the query builder can't understand it, rather than clearing it.

Thoughts?  I'm thinking I should do both of these things.